### PR TITLE
[Feature] Support SOLOv2

### PR DESCRIPTION
**Pull request by zhanggefan**
This PR adds support for SOLO-V2. The work is still in progress.
Special thanks to my colleague @Luhb for his kind help.
Special thanks to @WXinlong for his awesome work [mmdet-SOLO](https://github.com/WXinlong/SOLO)

- model part
  - [x] support solov2 model
  - [x] support solov2 light version
- configuration part
  - [x] solov2_r50_fpn_8gpu_1x_coco
  - [x] solov2_r50_fpn_8gpu_3x_coco
  - [x] solov2_r101_dcn_fpn_8gpu_3x_coco
  - [x] solov2_r101_fpn_8gpu_3x_coco
  - [x] solov2_x101_dcn_fpn_8gpu_3x_coco
  - [x] solov2_light_448_r18_fpn_8gpu_3x_coco
  - [x] solov2_light_448_r34_fpn_8gpu_3x_coco
  - [x] solov2_light_448_r50_fpn_8gpu_3x_coco
- docstring part
  - [x] model docstring
  - [x] configuration docstring
- performance alignment 

| Configuration  | Performance Here | Original |
| -- | -- | -- |
| solov2_r50_fpn_8gpu_1x_coco | [34.9](https://github.com/open-mmlab/mmdetection/files/8282239/20220315_124216.log.json.txt) | [34.8](https://github.com/WXinlong/SOLO/blob/95f3732d5fbb0d7c7044c7dd074f439d48a72ce5/README.md?plain=1#L50) |
| solov2_r50_fpn_8gpu_3x_coco | | |
| solov2_r101_dcn_fpn_8gpu_3x_coco | | |
| solov2_r101_fpn_8gpu_3x_coco | | |
| solov2_x101_dcn_fpn_8gpu_3x_coco | [42.5](https://github.com/open-mmlab/mmdetection/files/8355812/20220323_081035.log) | [42.4](https://github.com/WXinlong/SOLO/blob/95f3732d5fbb0d7c7044c7dd074f439d48a72ce5/README.md?plain=1#L54) |
| solov2_light_448_r18_fpn_8gpu_3x_coco | [29.7](https://github.com/open-mmlab/mmdetection/files/8300916/20220316_162541.log) | [29.6](https://github.com/WXinlong/SOLO/blob/95f3732d5fbb0d7c7044c7dd074f439d48a72ce5/README.md?plain=1#L62) |
| solov2_light_448_r34_fpn_8gpu_3x_coco | | |
| solov2_light_448_r50_fpn_8gpu_3x_coco | | |
| solov2_light_512_dcn_r50_fpn_8gpu_3x_coco | | |

### CODE REVIEW

File: mmdet/models/dense_heads/solo_head.py [hunk 1]
```diff
@@ -135,13 +135,18 @@ def resize_feats(self, feats):
         for i in range(len(feats)):
             if i == 0:
                 out.append(
-                    F.interpolate(feats[0], scale_factor=0.5, mode='bilinear'))
+                    F.interpolate(
+                        feats[0],
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/solo_head.py:139]: This may get some error if use scale_fctor. Because if the aaa cannot divide by 2, after resize, maybe it can't match the h and w?

File: mmdet/models/dense_heads/solo_head.py [hunk 2]
```diff
@@ -135,13 +135,18 @@ def resize_feats(self, feats):
...
+                        mode='bilinear',
+                        align_corners=False))
             elif i == len(feats) - 1:
                 out.append(
                     F.interpolate(
                         feats[i],
                         size=feats[i - 1].shape[-2:],
-                        mode='bilinear'))
+                        mode='bilinear',
+                        align_corners=False))
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/solo_head.py:149]: why delete this line?

File: mmdet/models/dense_heads/solov2_head.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+
+from mmdet.core import InstanceData, mask_matrix_nms, multi_apply
+from mmdet.core.utils import center_of_mass, generate_coordinate
+from mmdet.models.builder import HEADS
+from .solo_head import SOLOHead
+from mmcv.runner import BaseModule
+
+
+class MaskFeatModule(BaseModule):
+    def __init__(self,
```
Comment (BIGWangYuDong) [inline mmdet/models/dense_heads/solov2_head.py:17]: add doc strings

File: mmdet/models/dense_heads/solo_head.py (hunk 1 above)
Comment (zhanggefan) [reply mmdet/models/dense_heads/solo_head.py:139]: Yep. Personally, I think this change here in SOLO head actually makes little difference to the model's performance because there is no interaction across feature levels, and the code would work even when the feature map's sizes do not match across different levels. The only reason I did the change here is to make the logic looks better and nothing else ;-)

BTW, things are different in the code of MaskFeatureHead of SOLOv2. SOLOv2 uses unified MaskFeatureHead which uses element-add to fuse the features across levels. Thus the upsample and downsample output feature map shape must match the shape of the neighboring level.

File: mmdet/models/dense_heads/solo_head.py (hunk 2 above)
Comment (zhanggefan) [reply mmdet/models/dense_heads/solo_head.py:149]: I just add `align_corners=False` arg to suppress PyTorch's warning and `align_corners=False` is exactly the default behavior when this arg is not provided. 
But if we forgot to add this line, PyTorch would warn about the default behavior change at some old torch version.

File: configs/solo/solov2_light_r18_fpn_3x_coco.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        num_stages=4,
+        out_indices=(0, 1, 2, 3),
+        frozen_stages=1,
+        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet18'),
+        style='pytorch'),
+    neck=dict(
+        type='FPN',
+        in_channels=[64, 128, 256, 512],
+        out_channels=256,
+        start_level=0,
```
Comment (jshilong) [inline configs/solo/solov2_light_r18_fpn_3x_coco.py:21]: check this start_level

File: configs/solo/solov2_light_r18_fpn_3x_coco.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        nms_pre=500,
+        score_thr=0.1,
+        mask_thr=0.5,
+        filter_thr=0.05,
+        kernel='gaussian',  # gaussian/linear
+        sigma=2.0,
+        max_per_img=100))
+
+# runtime
+log_config = dict(
```
Comment (jshilong) [inline configs/solo/solov2_light_r18_fpn_3x_coco.py:60]: remove this

File: mmdet/models/dense_heads/solov2_head.py [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+        feat_channels (int): Number of hidden channels of the mask feature
+             map branch.
+        start_level (int): The starting feature map level from RPN that
+             will be used to predict the mask feature map. Default: 0.
+        end_level (int): The ending feature map level from rpn that
+             will be used to predict the mask feature map. Default: 3.
+        out_channels (int): Number of output channels of the mask feature
+             map branch. This is the channel count of the mask
+             feature map that to be dynamically convolved with the predicted
+             kernel. Default: 256.
```
Comment (jshilong) [inline mmdet/models/dense_heads/solov2_head.py:31]: no default value here

File: mmdet/models/dense_heads/solov2_head.py [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+            img_kernel_pred = torch.cat(img_kernel_pred, dim=0)
+            result = self._get_results_single(
+                img_kernel_pred,
+                img_cls_pred,
+                img_mask_feats,
+                img_meta=img_metas[img_id])
+            result_list.append(result)
+        return result_list
+
+    def _get_results_single(self,
```
Comment (jshilong) [inline mmdet/models/dense_heads/solov2_head.py:588]: Is it possible to reuse the code in SOLOHead?

File: configs/solo/solov2_light_r18_fpn_3x_coco.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+# data
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='LoadAnnotations', with_bbox=True, with_mask=True),
+    dict(
+        type='Resize',
+        img_scale=[(768, 512), (768, 480), (768, 448), (768, 416), (768, 384),
```
Comment (jshilong) [inline configs/solo/solov2_light_r18_fpn_3x_coco.py:86]: We recommend to re-arrange it as [(352, 768), ... (512, 768)]

File: configs/solo/solov2_x101_dcn_fpn_3x_coco.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
+_base_ = [
```
Comment (jshilong) [inline configs/solo/solov2_x101_dcn_fpn_3x_coco.py:1]: add configs/solo/solov2_r50_fpn_1x_coco.py to _base_ and remove duplicate parameters

File: mmdet/models/dense_heads/solov2_head.py [hunk 10]
```diff
@@ -0,0 +1,684 @@
...
+                            gt_bboxes,
+                            gt_labels,
+                            gt_masks,
+                            featmap_size=None):
+        device = gt_labels.device
+        gt_areas = torch.sqrt((gt_bboxes[:, 2] - gt_bboxes[:, 0]) *
+                              (gt_bboxes[:, 3] - gt_bboxes[:, 1]))
+
+        mlvl_mask_targets = []
+        mlvl_fg_pos = []
```
Comment (jshilong) [inline mmdet/models/dense_heads/solov2_head.py:298]: To be consistent, may we rename the `fg_pos` and `is_fg` as it in SOLOHead?

**Review by jshilong — COMMENTED** [review]
The PR is pretty good, but I recommend following the naming and logic of SOLOHead to make it more consistent

File: configs/solo/solov2_x101_dcn_fpn_3x_coco.py (hunk 9 above)
Comment (BIGWangYuDong) [reply configs/solo/solov2_x101_dcn_fpn_3x_coco.py:1]: @jshilong do we need to add new detector named `SOLOv2 `?

File: mmdet/models/dense_heads/solov2_head.py (hunk 7 above)
Comment (zhanggefan) [reply mmdet/models/dense_heads/solov2_head.py:588]: I actually tried but failed.

SOLOv2 uses dynamic convolution to decode the mask, and dynamic convolution cannot be done batch-wise, which means that we have to insert the code of mask decoding into a "do per-image" loop. Function `_get_result_single` is then a perfect candidate.

What's more, SOLO v1 does the score filtering inside the `_get_result_single`. if we reuse the code we would have nowhere to get the filtered indices, and thus we could not rule out the low-score dynamic convolution kernels before dynamic convolution. This would greatly harm the inference speed.

File: configs/solo/solov2_x101_dcn_fpn_3x_coco.py (hunk 9 above)
Comment (zhanggefan) [reply configs/solo/solov2_x101_dcn_fpn_3x_coco.py:1]: @jshilong great suggestion. I will change it.

Comment (zhanggefan) [reply configs/solo/solov2_x101_dcn_fpn_3x_coco.py:1]: @BIGWangYuDong Is it really necessary to add one more pure alias class?

File: mmdet/models/dense_heads/solov2_head.py (hunk 10 above)
Comment (zhanggefan) [reply mmdet/models/dense_heads/solov2_head.py]: Lots of variables have been renamed by now. 

To be honest, I hate touching those codes for a second time... target assigning code is inherently complicated for the SOLO series. Renaming those variables does not make the logic clearer either... The only way to understand it thoroughly is by step-by-step debugging. 

Because the renaming introduces lots of changes, I cannot guarantee the correctness and that the performance is not affected. Please help me redo the performance alignment training because I do not have the resources for it now.

File: mmdet/models/dense_heads/solov2_head.py (hunk 6 above)
Comment (zhanggefan) [reply mmdet/models/dense_heads/solov2_head.py]: fixed

File: configs/solo/solov2_light_r18_fpn_3x_coco.py (hunk 5 above)
Comment (zhanggefan) [reply configs/solo/solov2_light_r18_fpn_3x_coco.py]: fixed

**Comment by jshilong** [comment]
@BIGWangYuDong Would you mind doing the renaming and helping to benchmark the pr? The PR  LGTM except for the variable naming is inconsistent with SOLOv1.

File: mmdet/models/dense_heads/solov2_head.py [hunk 11]
```diff
@@ -0,0 +1,687 @@
...
+            if i == 0:
+                one_conv = ConvModule(
+                    self.in_channels,
+                    self.feat_channels,
+                    3,
+                    padding=1,
+                    conv_cfg=self.conv_cfg,
+                    norm_cfg=self.norm_cfg,
+                    inplace=False)
+                convs_per_level.add_module('conv' + str(i), one_conv)
```
Comment (jshilong) [inline mmdet/models/dense_heads/solov2_head.py:69]: @BIGWangYuDong Please help to rename the module name in the head to align with other heads

**Review by jshilong — COMMENTED** [review]
@BIGWangYuDong The Conv module names should also keep the same style with other heads

**Comment by BIGWangYuDong** [comment]
| Configuration  | Performance Here | Original |
| -- | -- | -- |
| solov2_r50_fpn_8gpu_1x_coco |    34.8                    |34.8 |
| solov2_r50_fpn_8gpu_3x_coco |       37.6                    |37.5 |
| solov2_light_448_r18_fpn_8gpu_3x_coco | 29.7          |29.6 |
| solov2_light_448_r34_fpn_8gpu_3x_coco | 31.8         |32.0 |
| solov2_light_448_r50_fpn_8gpu_3x_coco | 33.6      |33.7 |

File: mmdet/models/dense_heads/solov2_head.py (hunk 7 above)
Comment (BIGWangYuDong) [reply mmdet/models/dense_heads/solov2_head.py:591]: There are some differences between v1 and v2, it is hard to reuse SOLOHead.

**Comment by BIGWangYuDong** [comment]
kindly ping @zhanggefan @jshilong have a look

**Comment by BIGWangYuDong** [comment]
TODO:
- [ ] Prepare models
- [ ] Add README in config files

**Review by jshilong — APPROVED** [review]
LGTM, can be merged after adding unitests

**Comment by ZwwWayne** [comment]
Overall, this PR looks great. It could be better if unit tests could be add.

**Comment by ZwwWayne** [comment]
can be merged after resolving conflicts

**Comment by BIGWangYuDong** [comment]
> can be merged after resolving conflicts

Done

**Review by ZwwWayne — APPROVED** [review]

**Comment by qihao-plus** [comment]
can't download SOLOv2 model from https://download.openmmlab.com/mmdetection/v2.0/solov2/solov2_r50_fpn_1x_coco/solov2_r50_fpn_1x_coco_20220512_125858-a357fa23.pth

and https://github.com/open-mmlab/mmdetection/blob/dev/configs/solov2/metafile.yml
